File: src/main/java/com/dannbrown/musicbox/client/ClientAudioManager.java
package com.dannbrown.musicbox.client;

import com.dannbrown.musicbox.client.exe.YoutubeDL;
import com.dannbrown.musicbox.main.MusicBox;
import com.dannbrown.musicbox.main.YoutubeUtils;

import java.io.*;
import java.util.concurrent.CompletableFuture;

public class ClientAudioManager {
  public static CompletableFuture<Boolean> downloadAudio(String url, String fileName) {
    if (MusicBox.PATH == null) {
      return CompletableFuture.completedFuture(false);
    }
    return CompletableFuture.supplyAsync(() -> YoutubeDL.executeCommand(
            YoutubeUtils.removeUrlParameters(url),
      "-x",
      "--no-progress",
      "--no-playlist",
      "--external-downloader-args", String.format("\"-ss 00:00:00.00 -to 00:%s:00.00\"", MusicBox.MAX_AUDIO_MINUTES),
      "--break-match-filter", "ext~=3gp|aac|flv|m4a|mov|mp3|mp4|ogg|wav|webm",
      "--audio-format", "vorbis",
      "--audio-quality", MusicBox.AUDIO_BIT_RATE.getValue(),
      "--postprocessor-args", String.format("ffmpeg:-ac 1 -t %s", MusicBox.MAX_AUDIO_MINUTES * 60),
      "--ffmpeg-location", MusicBox.PATH.resolve("ffmpeg").toString(),
      "-o", fileNameToFile(fileName).toString())
    );
  }

  public static InputStream getAudioInputStream(String fileName) {
    try {
      File file = fileNameToFile(fileName);
      return new FileInputStream(file);
    } catch (FileNotFoundException e) {
      return null;
    }
  }

  public static File fileNameToFile(String fileName) {
    if (MusicBox.PATH == null) {
      return null;
    }
    return new File(MusicBox.PATH.resolve("client_downloads/" + fileName).toString());
  }
}

File: src/main/kotlin/com/dannbrown/musicbox/lib/client/ClientAudioManager.kt
package com.dannbrown.musicbox.lib.client

import com.dannbrown.musicbox.lib.executable.YoutubeDL
import com.dannbrown.musicbox.lib.main.Manager
import com.dannbrown.musicbox.lib.main.YoutubeUtils
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream
import java.util.concurrent.CompletableFuture

object ClientAudioManager {
  fun downloadAudio(url: String, fileName: String): CompletableFuture<Boolean> {
    if (Manager.PATH == null) {
      return CompletableFuture.completedFuture(false)
    }
    return CompletableFuture.supplyAsync {
      YoutubeDL.executeCommand(
        YoutubeUtils.removeUrlParameters(url),
        "-x",
        "--no-progress",
        "--no-playlist",
        "--external-downloader-args",
        String.format("\"-ss 00:00:00.00 -to 00:%s:00.00\"", Manager.MAX_AUDIO_MINUTES),
        "--break-match-filter",
        "ext~=3gp|aac|flv|m4a|mov|mp3|mp4|ogg|wav|webm",
        "--audio-format",
        "vorbis",
        "--audio-quality",
        Manager.AUDIO_BIT_RATE.value,
        "--postprocessor-args",
        String.format("ffmpeg:-ac 1 -t %s", Manager.MAX_AUDIO_MINUTES * 60),
        "--ffmpeg-location",
        Manager.PATH!!.resolve("ffmpeg")
          .toString(),
        "-o",
        fileNameToFile(fileName)
          .toString())
    }
  }

  @JvmStatic
  fun getAudioInputStream(fileName: String): InputStream? {
    try {
      val file = fileNameToFile(fileName)
      return FileInputStream(file)
    } catch (e: FileNotFoundException) {
      return null
    }
  }

  fun fileNameToFile(fileName: String): File? {
    if (Manager.PATH == null) {
      return null
    }
    return File(Manager.PATH!!.resolve("client_downloads/$fileName")
      .toString())
  }
}
